File: src/main/java/Electro2D/DotThread.java
package main.java.Electro2D;

import javajs.async.SwingJSUtils;
import javajs.async.SwingJSUtils.StateHelper;
import javajs.async.SwingJSUtils.StateMachine;

/**
 * Electro2D.DotThread is the thread controlling the second half of the 2DGE animation.
 *
 * @author Jill Zapoticznyj
 * @author Adam Bazinet
 */

public class DotThread extends Thread implements StateMachine {
    private static boolean go = false;
    private GelCanvas gel;
    private Electro2D electro2D;
	private StateHelper stateHelper;
	private int loop_i;
	private long tm;

    /**
     * Constructor
     *
     * @param e reference to the Electro2D
     */

    public DotThread(Electro2D e) {
        gel = e.getGel();
        electro2D = e;
    }

    /**
     * This method called when the play button is pressed to pause
     * the animation or when the stop button is pressed.  Sets the
     * boolean variable in charge of the animation loop to false,
     * stopping the animation.
     */

    public void stopDots() {
        go = false;
    }

    /**
     * This method returns the value of go - the variable
     * which determines whether or not to edit the location of the protein
     * dots.
     */

    public static boolean getDotState() {
        return go;
    }

    /**
     * This method is called just before the animation is started or
     * when the animation is restarted by pushing the play button.
     */

    public void startDots() {
        go = true;
    }


    /**
     * The run method used by the animation thread.  Makes repeated
     * calls to Electro2D.GelCanvas' genDots method while go is true.
     */

    private final static int INIT = 0;
    private final static int LOOP_IEF = 1;
    private final static int LOOP_DOTS = 2;
    
    
    public void run() {
    	stateHelper = new SwingJSUtils.StateHelper(this);
    	stateHelper.next(INIT);
    }

	@Override
	public boolean stateLoop() {
		if (stateHelper.isAlive()) {
			switch (stateHelper.getState()) {
			case INIT:
				GelCanvas.stopBlink();
				// send the percent acrylamide value to Electro2D.ProteinDot
				ProteinDot.setPercent(electro2D.getLowPercent(), electro2D.getHighPercent());

				gel.resetReLine();
				stateHelper.next(LOOP_IEF);
				break;
			case LOOP_IEF:
				// while the IEFProteins are still visible...
				// LOOP_IEF
				if (IEFProtein.returnHeight() > 0) {
					// ...shrink them in size...
					IEFProtein.shrinkProtein();
					// ...and redraw them to the Electro2D.GelCanvas
					gel.shrinkIEF();
					// then wait for 100 milliseconds
					stateHelper.sleep(100);
					break;
				}
				// Make the ProteinDots visible
				if (ProteinDot.getShow() == false) {
					ProteinDot.setShow();
				}
				loop_i = 0;
				tm = System.currentTimeMillis();
				stateHelper.next(LOOP_DOTS);
				break;
			case LOOP_DOTS:
				// While the stop button or pause button has not been pressed...
				if (go) {
					// ...change the location of the dots and redraw them to
					// the Electro2D.GelCanvas
					gel.genDots();
					if (System.currentTimeMillis() - tm >= 10000) {
						stopDots();
					}
					loop_i = 1 + loop_i;
					stateHelper.sleep(100);
					break;
				}
				gel.setreLine();
				gel.setMWLines(loop_i);
				loop_i = 0;
				electro2D.resetPlay();
				//gel.paint(gel.getGraphic());
				gel.repaint();
				break;
			}
		}
		return false;
	}
}

File: site/swingjs/j2s/main/java/Electro2D/DotThread.js
(function(){var P$=Clazz.newPackage("main.java.Electro2D"),I$=[[0,['javajs.async.SwingJSUtils','.StateHelper'],'main.java.Electro2D.GelCanvas','main.java.Electro2D.ProteinDot','main.java.Electro2D.IEFProtein']],I$0=I$[0],$I$=function(i,n,m){return m?$I$(i)[n].apply(null,m):((i=(I$[i]||(I$[i]=Clazz.load(I$0[i])))),!n&&i.$load$&&Clazz.load(i,2),i)};
/*c*/var C$=Clazz.newClass(P$, "DotThread", null, 'Thread', [['javajs.async.SwingJSUtils','javajs.async.SwingJSUtils.StateMachine']]);

C$.$clinit$=2;

Clazz.newMeth(C$, '$init$', function () {
this.delay=10 ||100;
},1);

C$.$fields$=[['I',['loop_i','delay'],'J',['tm'],'O',['gel','main.java.Electro2D.GelCanvas','electro2D','main.java.Electro2D.Electro2D','stateHelper','javajs.async.SwingJSUtils.StateHelper']]
,['Z',['go']]]

Clazz.newMeth(C$, 'c$$main_java_Electro2D_Electro2D',  function (e) {
Clazz.super_(C$, this);
this.gel=e.getGel$();
this.electro2D=e;
}, 1);

Clazz.newMeth(C$, 'stopDots$',  function () {
C$.go=false;
});

Clazz.newMeth(C$, 'getDotState$',  function () {
return C$.go;
}, 1);

Clazz.newMeth(C$, 'startDots$',  function () {
C$.go=true;
});

Clazz.newMeth(C$, 'run$',  function () {
this.stateHelper=Clazz.new_($I$(1,1).c$$javajs_async_SwingJSUtils_StateMachine,[this]);
this.stateHelper.next$I(0);
});

Clazz.newMeth(C$, 'stateLoop$',  function () {
if (this.stateHelper.isAlive$()) {
switch (this.stateHelper.getState$()) {
case 0:
$I$(2).stopBlink$();
$I$(3,"setPercent$D$D",[this.electro2D.getLowPercent$(), this.electro2D.getHighPercent$()]);
this.gel.setRedrawPHLines$Z(false);
this.stateHelper.next$I(1);
break;
case 1:
if ($I$(4).returnHeight$() > 0) {
$I$(4).shrinkProtein$();
this.gel.shrinkIEF$();
this.stateHelper.sleep$I(this.delay);
break;
}if ($I$(3).getShow$() == false ) {
$I$(3).setShow$();
}this.loop_i=0;
this.tm=System.currentTimeMillis$();
this.stateHelper.next$I(2);
break;
case 2:
if (C$.go) {
this.gel.genDots$();
if (Long.$ge(Long.$sub(System.currentTimeMillis$(),this.tm),10000 )) {
this.stopDots$();
}this.loop_i=1 + this.loop_i;
this.stateHelper.sleep$I(this.delay);
break;
}this.gel.setRedrawPHLines$Z(true);
this.gel.setMWLines$I(this.loop_i);
this.loop_i=0;
this.electro2D.resetPlay$();
this.gel.repaint$();
break;
}
}return false;
});

C$.$static$=function(){C$.$static$=0;
C$.go=false;
};

Clazz.newMeth(C$);
})();
;Clazz.setTVer('3.3.1-v4');//Created 2022-04-21 17:00:07 Java2ScriptVisitor version 3.3.1-v4 net.sf.j2s.core.jar version 3.3.1-v4
